File: C++/G04.cpp
#include <iostream>
#include <vector>
#include <stack>
using namespace std;

/**
 * Depth First Search on Undirected Graph 
 * 
 * Note : DFS uses stack while BFS uses queue.
 * 
 * Input Format :
 * first line contains 2 space seperated integers n (number of nodes) and m (number of edges).
 * each of next m lines contains u (starting node of edge) and v (ending node of edge).
 * 
 * Time Complexity  : O(n) 
 * 
 * Space Complexity : O(n)   +   O(n)   +   O(n)  ≈ O(n) 
 *                     ↑          ↑          ↑
 *                   stack  dfs_traversal visited  
 * 
 * Reference : https://youtu.be/uDWljP2PGmU
 **/

vector<int> dfs(int n, vector<int> adj_list[]) {
    vector<int> dfs_traversal;
    vector<int> visited(n, 0); 
    for (int i = 0; i < n; ++i) {
        if (!visited[i]) {
            stack<int> s;
            s.push(i);
            visited[i] = 1;
            while(!s.empty()) {
                int node = s.top();
                s.pop();
                dfs_traversal.emplace_back(node);
                for (int& k : adj_list[node]) {
                    if (!visited[k]) {
                        s.push(k);
                        visited[k] = 1;
                    }
                }
            }
        }
    }
    return dfs_traversal;
}

int main() {
    int n, m;
    cin >> n >> m;

    vector<int> adj_list[n];

    for (int i = 0; i < m; ++i) {
        int u, v;
        cin >> u >> v;
        adj_list[u - 1].emplace_back(v - 1);
        adj_list[v - 1].emplace_back(u - 1);
    }

    for (int i = 0; i < n; ++i) {
        cout << i + 1 << " -> ";
        for (int& j : adj_list[i]) {
            cout << j + 1 << " ";
        }
        cout << "\n";
    }

    vector<int> dfs_traversal = dfs(n, adj_list);
    cout << "DFS : ";
    for (int& i : dfs_traversal) {
        cout << i + 1 << " ";
    }
    cout << "\n";
}

/**
 * Input :
 * 7 6
 * 1 2
 * 2 3
 * 2 7
 * 3 5
 * 4 6
 * 5 7
 * 
 * Output :
 * 1 -> 2 
 * 2 -> 1 3 7 
 * 3 -> 2 5 
 * 4 -> 6 
 * 5 -> 3 7 
 * 6 -> 4 
 * 7 -> 2 5
 * DFS : 1 2 7 5 3 4 6
 **/

File: Python/G04.py
from collections import deque as stack
from typing import List

# Depth First Search on Undirected Graph 
# 
# Input Format :
# first line contains 2 space seperated integers n (number of nodes) and m (number of edges).
# each of next m lines contains u (starting node of edge) and v (ending node of edge).
# 
# Time Complexity  : O(n) 
# 
# Space Complexity : O(n)   +   O(n)   +   O(n)  ≈ O(n) 
#                     ↑          ↑          ↑
#                   stack  dfs_traversal visited 
# 
# Reference : https://youtu.be/uDWljP2PGmU

def dfs(n: int, adj_list: List[int]) -> List[int]:
    dfs_traversal = []
    visited = [False for i in range(n)] 
    for i in range(n):
        if not visited[i]:
            s = stack()
            s.append(i) # push
            visited[i] = True
            while s: # q not empty
                node = s.pop() # top
                dfs_traversal.append(node)
                for k in adj_list[node]:
                    if not visited[k]:
                        s.append(k)
                        visited[k] = True

    return dfs_traversal

if __name__ == "__main__":

    n, m = [int(x) for x in input().split()]
    adj_list = [[] for i in range(n)]

    for i in range(m):
        u, v = [int(x) for x in input().split()]
        adj_list[u - 1].append(v - 1)
        adj_list[v - 1].append(u - 1)

    for i in range(n):
        print(f"{i + 1} -> ", end="")
        for j in adj_list[i]:
            print(f"{j + 1} ", end="")
        print()

    dfs_traversal = dfs(n, adj_list)
    print("DFS : ", end="")
    for i in dfs_traversal:
        print(f"{i + 1} ", end="")
    print()

# Input :
# 7 6
# 1 2
# 2 3
# 2 7
# 3 5
# 4 6
# 5 7
# 
# Output :
# 1 -> 2 
# 2 -> 1 3 7 
# 3 -> 2 5 
# 4 -> 6 
# 5 -> 3 7 
# 6 -> 4 
# 7 -> 2 5
# DFS : 1 2 7 5 3 4 6 